apiserver/common/modelmanagerinterface.go

```diff
@@ -111,12 +112,21 @@ type modelManagerStateShim struct {
 	*state.State
 	model *state.Model
 	pool  *state.StatePool
+	user  names.UserTag
 }
 
 // NewModelManagerBackend returns a modelManagerStateShim wrapping the passed
 // state, which implements ModelManagerBackend.
 func NewModelManagerBackend(m *state.Model, pool *state.StatePool) ModelManagerBackend {
-	return modelManagerStateShim{m.State(), m, pool}
+	return modelManagerStateShim{m.State(), m, pool, names.UserTag{}}
+}
+
+// NewUserAwareModelManagerBackend returns a user-aware modelManagerStateShim
+// wrapping the passed state, which implements ModelManagerBackend. The
+// returned backend may emit redirect errors when attempting a model lookup for
+// a migrated model that this user had been granted access to.
+func NewUserAwareModelManagerBackend(m *state.Model, pool *state.StatePool, u names.UserTag) ModelManagerBackend {
```

Comment:
I was going to ask why don't we handle this a model login time, but this is for the controller isn't it?
Is the controller login still accessed through the `/api` root?

Reply:
Any call that logins to the migrated model will already get the redirect error from the changes introduced via previous PRs. However, things like `juju show-model` login to the controller first and then make an RPC (e.g `ModelInfo`) where they pass the model that may have been potentially migrated. 

So basically, [here](https://github.com/juju/juju/blob/develop/apiserver/facades/client/modelmanager/modelmanager.go#L184-L185) both lines point to the same _controller_ model and the code calls `GetBackend` on the model manager interface from L184 to obtain a new interface for the model UUIDs passed to `ModelInfo`.